File: go/monocr.go
package monocr

import (
	_ "embed"
	"fmt"
	"image"
	_ "image/jpeg"
	_ "image/png"
	"os"
	"os/exec"
	"path/filepath"
	"strings"

	"github.com/MonDevHub/monocr-onnx/go/pkg/model"
	"github.com/MonDevHub/monocr-onnx/go/pkg/predictor"
)

//go:embed charset.txt
var embeddedCharset string

// ReadImage recognizes text from an image file.
// It automatically downloads the model if not present.
func ReadImage(imagePath string) (string, error) {
	manager, err := model.NewManager()
	if err != nil {
		return "", err
	}

	modelPath, err := manager.GetModelPath()
	if err != nil {
		return "", err
	}

	return ReadImageWithModel(imagePath, modelPath, embeddedCharset)
}

// ReadImages recognizes text from multiple image files.
func ReadImages(imagePaths []string) ([]string, error) {
	manager, err := model.NewManager()
	if err != nil {
		return nil, err
	}

	modelPath, err := manager.GetModelPath()
	if err != nil {
		return nil, err
	}

	pred, err := predictor.NewPredictor(modelPath, embeddedCharset)
	if err != nil {
		return nil, err
	}
	defer pred.Close()

	var results []string
	for _, path := range imagePaths {
		text, err := predictFile(pred, path)
		if err != nil {
			return nil, err
		}
		results = append(results, text)
	}
	return results, nil
}

// ReadImageWithAccuracy recognizes text and calculates accuracy against ground truth.
func ReadImageWithAccuracy(imagePath, groundTruth string) (string, float64, error) {
	text, err := ReadImage(imagePath)
	if err != nil {
		return "", 0, err
	}
	accuracy := calculateAccuracy(text, groundTruth)
	return text, accuracy, nil
}

// ReadImageWithModel allows specifying custom model and charset paths.
func ReadImageWithModel(imagePath, modelPath, charset string) (string, error) {
	pred, err := predictor.NewPredictor(modelPath, charset)
	if err != nil {
		return "", err
	}
	defer pred.Close()

	return predictFile(pred, imagePath)
}

func predictFile(pred *predictor.Predictor, imagePath string) (string, error) {
	f, err := os.Open(imagePath)
	if err != nil {
		return "", err
	}
	defer f.Close()

	img, _, err := image.Decode(f)
	if err != nil {
		return "", fmt.Errorf("failed to decode image: %v", err)
	}

	return pred.Predict(img)
}

// ReadPDF recognizes text from a PDF file (requires pdftoppm/poppler-utils).
func ReadPDF(pdfPath string) ([]string, error) {
	// Check for pdftoppm
	_, err := exec.LookPath("pdftoppm")
	if err != nil {
		return nil, fmt.Errorf("pdftoppm not found: please install poppler-utils")
	}

	manager, err := model.NewManager()
	if err != nil {
		return nil, err
	}

	modelPath, err := manager.GetModelPath()
	if err != nil {
		return nil, err
	}

	return readPDFWithModel(pdfPath, modelPath, embeddedCharset)
}

// ReadPDFs recognizes text from multiple PDF files.
func ReadPDFs(pdfPaths []string) ([][]string, error) {
	// Check for pdftoppm
	_, err := exec.LookPath("pdftoppm")
	if err != nil {
		return nil, fmt.Errorf("pdftoppm not found: please install poppler-utils")
	}

	manager, err := model.NewManager()
	if err != nil {
		return nil, err
	}

	modelPath, err := manager.GetModelPath()
	if err != nil {
		return nil, err
	}

	var results [][]string
	for _, path := range pdfPaths {
		pages, err := readPDFWithModel(path, modelPath, embeddedCharset)
		if err != nil {
			return nil, err
		}
		results = append(results, pages)
	}
	return results, nil
}

func readPDFWithModel(pdfPath, modelPath, charset string) ([]string, error) {
	// Create temp dir
	tempDir, err := os.MkdirTemp("", "monocr-go-")
	if err != nil {
		return nil, err
	}
	defer os.RemoveAll(tempDir)

	// Convert PDF to images
	cmd := exec.Command("pdftoppm", "-png", "-r", "300", pdfPath, filepath.Join(tempDir, "page"))
	if err := cmd.Run(); err != nil {
		return nil, fmt.Errorf("failed to convert PDF: %v", err)
	}

	// Read all generated images
	files, err := os.ReadDir(tempDir)
	if err != nil {
		return nil, err
	}

	pred, err := predictor.NewPredictor(modelPath, charset)
	if err != nil {
		return nil, err
	}
	defer pred.Close()

	var results []string
	for _, file := range files {
		if strings.HasSuffix(file.Name(), ".png") {
			imgPath := filepath.Join(tempDir, file.Name())
			text, err := predictFile(pred, imgPath)
			if err != nil {
				continue
			}
			results = append(results, text)
		}
	}

	return results, nil
}

// Levenshtein distance calculation
func levenshtein(s1, s2 []rune) int {
	len1, len2 := len(s1), len(s2)
	column := make([]int, len1+1)

	for y := 1; y <= len1; y++ {
		column[y] = y
	}

	for x := 1; x <= len2; x++ {
		column[0] = x
		lastDiag := x - 1
		for y := 1; y <= len1; y++ {
			oldDiag := column[y]
			cost := 0
			if s1[y-1] != s2[x-1] {
				cost = 1
			}
			column[y] = min(column[y]+1, min(column[y-1]+1, lastDiag+cost))
			lastDiag = oldDiag
		}
	}
	return column[len1]
}

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}

func calculateAccuracy(pred, truth string) float64 {
	p := []rune(pred)
	t := []rune(truth)
	
	if len(t) == 0 {
		if len(p) == 0 {
			return 100.0
		}
		return 0.0
	}
	
	dist := levenshtein(p, t)
	maxLen := len(p)
	if len(t) > maxLen {
		maxLen = len(t)
	}
	
	if maxLen == 0 {
		return 100.0
	}
	
	return (1.0 - float64(dist)/float64(maxLen)) * 100.0
}

File: js/bin/monocr.js
#!/usr/bin/env node

const { program } = require('commander');
const { read_image, read_pdf } = require('../src/index');
const fs = require('fs');
const path = require('path');

program
    .name('monocr')
    .description('Mon language OCR using ONNX Runtime')
    .version('0.1.0');

program
    .command('image <path>')
    .description('Recognize text from an image file')
    .option('-m, --model <path>', 'Path to ONNX model (optional, auto-downloads)')
    .option('-c, --charset <path>', 'Path to charset file (optional)')
    .action(async (imagePath, options) => {
        try {
            const text = await read_image(imagePath, options.model, options.charset);
            console.log(text);
        } catch (err) {
            console.error('Error:', err.message);
            process.exit(1);
        }
    });

program
    .command('pdf <path>')
    .description('Recognize text from a PDF file')
    .option('-m, --model <path>', 'Path to ONNX model (optional, auto-downloads)')
    .option('-c, --charset <path>', 'Path to charset file (optional)')
    .action(async (pdfPath, options) => {
        try {
            const pages = await read_pdf(pdfPath, options.model, options.charset);
            pages.forEach((pageText, i) => {
                console.log(`--- Page ${i + 1} ---`);
                console.log(pageText);
                console.log();
            });
        } catch (err) {
            console.error('Error:', err.message);
            process.exit(1);
        }
    });

program
    .command('batch <directory>')
    .description('Process all images in a directory')
    .option('-m, --model <path>', 'Path to ONNX model (optional, auto-downloads)')
    .option('-c, --charset <path>', 'Path to charset file (optional)')
    .option('-o, --output <path>', 'Output file for results (optional)')
    .action(async (directory, options) => {
        try {
            const files = fs.readdirSync(directory)
                .filter(f => /\.(jpg|jpeg|png)$/i.test(f))
                .sort();
            
            const results = [];
            
            for (const file of files) {
                const filePath = path.join(directory, file);
                console.error(`Processing: ${file}...`);
                
                try {
                    const text = await read_image(filePath, options.model, options.charset);
                    results.push({ file, text, success: true });
                } catch (err) {
                    results.push({ file, error: err.message, success: false });
                }
            }
            
            if (options.output) {
                fs.writeFileSync(options.output, JSON.stringify(results, null, 2));
                console.error(`Results written to ${options.output}`);
            } else {
                console.log(JSON.stringify(results, null, 2));
            }
        } catch (err) {
            console.error('Error:', err.message);
            process.exit(1);
        }
    });

program
    .command('download')
    .description('Download model files to local cache')
    .action(async () => {
        try {
            const { MonOCR } = require('../src/index');
            const ocr = new MonOCR();
            await ocr.modelManager.downloadModel();
        } catch (err) {
            console.error('Error:', err.message);
            process.exit(1);
        }
    });

program.parse();
